stratum/server.go

```diff
@@ -0,0 +1,2262 @@
+		s.sessionsSHA[sess] = struct{}{}
+	case "scrypt":
+		s.sessionsScrypt[sess] = struct{}{}
+	case "kawpow":
+		s.sessionsKawpow[sess] = struct{}{}
+	}
+}
+
+// unregisterSession removes a session from the appropriate algorithm's session map
+func (s *Server) unregisterSession(sess *session) {
+	s.sessionsMu.Lock()
+	defer s.sessionsMu.Unlock()
+	switch sess.chain {
+	case "sha256":
+		delete(s.sessionsSHA, sess)
+	case "scrypt":
+		delete(s.sessionsScrypt, sess)
+	case "kawpow":
+		delete(s.sessionsKawpow, sess)
+	}
+}
+
+func (s *Server) Start() error {
+	if s.backend == nil {
+		return fmt.Errorf("nil backend")
+	}
+
+	// Start the job worker pool
+	s.jobPool.start()
+
+	// Start SHA listener if configured
+	if s.config.SHAAddr != "" {
+		ln, err := net.Listen("tcp", s.config.SHAAddr)
+		if err != nil {
+			return fmt.Errorf("failed to start SHA listener on %s: %v", s.config.SHAAddr, err)
+		}
+		s.lnSHA = ln
+		s.logger.WithField("addr", s.config.SHAAddr).Info("sha256 stratum listener started")
+		go s.acceptLoop(ln, "sha256")
+	}
+
+	// Start Scrypt listener if configured
+	if s.config.ScryptAddr != "" {
+		ln, err := net.Listen("tcp", s.config.ScryptAddr)
+		if err != nil {
+			return fmt.Errorf("failed to start Scrypt listener on %s: %v", s.config.ScryptAddr, err)
+		}
+		s.lnScrypt = ln
+		s.logger.WithField("addr", s.config.ScryptAddr).Info("Scrypt stratum listener started")
+		go s.acceptLoop(ln, "scrypt")
+	}
+
+	// Start Kawpow listener if configured
+	if s.config.KawpowAddr != "" {
+		ln, err := net.Listen("tcp", s.config.KawpowAddr)
+		if err != nil {
+			return fmt.Errorf("failed to start Kawpow listener on %s: %v", s.config.KawpowAddr, err)
+		}
+		s.lnKawpow = ln
+		s.logger.WithField("addr", s.config.KawpowAddr).Info("Kawpow stratum listener started")
+		go s.acceptLoop(ln, "kawpow")
+	}
+
+	// Start template polling loops for each algorithm (check every second)
+	if s.config.SHAAddr != "" {
+		s.wg.Add(1)
+		go s.templatePollingLoop("sha256")
+	}
+	if s.config.ScryptAddr != "" {
+		s.wg.Add(1)
+		go s.templatePollingLoop("scrypt")
+	}
+	if s.config.KawpowAddr != "" {
+		s.wg.Add(1)
+		go s.templatePollingLoop("kawpow")
+	}
+
+	// Start force broadcast loops to ensure miners always have work.
+	// Runs every 1 second (minimum frequency granularity) and checks each session's
+	// individual jobFrequency setting. Miners can configure their frequency via username:
+	// address.workername.frequency=X (1-10 seconds, supports decimals, default 5s)
+	if s.config.SHAAddr != "" {
+		s.wg.Add(1)
+		go s.forceBroadcastLoop("sha256")
+	}
+	if s.config.ScryptAddr != "" {
+		s.wg.Add(1)
+		go s.forceBroadcastLoop("scrypt")
+	}
+	if s.config.KawpowAddr != "" {
+		s.wg.Add(1)
+		go s.forceBroadcastLoop("kawpow")
+	}
+
+	return nil
+}
+
+// templatePollingLoop polls for new templates and broadcasts to connected miners
+func (s *Server) templatePollingLoop(algorithm string) {
+	defer s.wg.Done()
+	ticker := time.NewTicker(1 * time.Second)
```

Comment:
we can make this a constant would be easy to track and change

Reply:
Fixed